### PR TITLE
`-Znext-solver` caching

### PR DESCRIPTION
**Pull request by lcnr**
This PR has two major changes while also fixing multiple issues found via fuzzing.

The main optimization is the ability to not discard provisional cache entries when popping the highest cycle head the entry depends on. This fixes the hang in Fuchsia with `-Znext-solver=coherence`.

It also bails if the result of a fixpoint iteration is ambiguous, even without reaching a fixpoint. This is necessary to avoid exponential blowup if a coinductive cycle results in ambiguity, e.g. due to unknowable candidates in coherence.

Updating stack entries pretty much exclusively happens lazily now, so `fn check_invariants` ended up being mostly useless and I've removed it. See https://gist.github.com/lcnr/8de338fdb2685581e17727bbfab0622a for the invariants we would be able to assert with it.

For a general overview, see the in-process update of the relevant rustc-dev-guide chapter: https://hackmd.io/1ALkSjKlSCyQG-dVb_PUHw

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_type_ir/src/search_graph/mod.rs [hunk 1]
```diff
@@ -273,67 +384,62 @@ impl<D: Delegate<Cx = X>, X: Cx> SearchGraph<D> {
...
+        if let Some(parent_index) = stack.last_index() {
+            let parent = &mut stack[parent_index];
             parent.reached_depth = parent.reached_depth.max(reached_depth);
             parent.encountered_overflow |= encountered_overflow;
+
+            parent.heads.extend_from_child(parent_index, heads);
+            let step_kind = Self::step_kind(cx, parent.input);
+            parent.nested_goals.extend_from_child(step_kind, nested_goals);
+            if !nested_goals.is_empty() {
+                parent.nested_goals.insert(parent.input, UsageKind::Single(PathKind::Coinductive))
```
Comment (compiler-errors) [inline compiler/rustc_type_ir/src/search_graph/mod.rs:408]: What is this for?

File: compiler/rustc_type_ir/src/search_graph/mod.rs [hunk 2]
```diff
@@ -538,55 +934,83 @@ impl<D: Delegate<Cx = X>, X: Cx> SearchGraph<D> {
...
+            //
+            // See tests/ui/traits/next-solver/cycles/fixpoint-rerun-all-cycle-heads.rs
+            // for an example.
+            //
+            // Check whether we reached a fixpoint, either because the final result
+            // is equal to the provisional result of the previous iteration, or because
+            // this was only the root of either coinductive or inductive cycles, and the
+            // final result is equal to the initial response for that case.
+            if self.reached_fixpoint(cx, &stack_entry, usage_kind, result) {
+                self.rebase_provisional_cache_entries(cx, &stack_entry, |_, result| result);
```
Comment (compiler-errors) [inline compiler/rustc_type_ir/src/search_graph/mod.rs:969]: Did you decide against the `rotate_*` terminology?

File: compiler/rustc_type_ir/src/search_graph/mod.rs [hunk 3]
```diff
@@ -403,57 +518,319 @@ impl<D: Delegate<Cx = X>, X: Cx> SearchGraph<D> {
...
+    /// head from the stack after we've finished computing it, we can't be sure that the
+    /// provisional cache entry is still applicable. We need to keep the cache entries to
+    /// prevent hangs.
+    ///
+    /// What we therefore do is check whether the cycle kind of all cycles the goal of a
+    /// provisional cache entry is involved in would stay the same when computing the
+    /// goal without its cycle head on the stack. For more details, see the relevant
+    /// [rustc-dev-guide chapter](https://rustc-dev-guide.rust-lang.org/solve/caching.html).
+    ///
+    /// This can be thought of rotating the sub-tree of this provisional result and changing
```
Comment (compiler-errors) [inline compiler/rustc_type_ir/src/search_graph/mod.rs:609]: Ah, you left a comment here saying it's like rotating. That's fine.

**Review by compiler-errors — COMMENTED** [review]
One nit

File: compiler/rustc_type_ir/src/search_graph/mod.rs (hunk 1 above)
Comment (lcnr) [reply compiler/rustc_type_ir/src/search_graph/mod.rs:408]: good catch. deserves a comment

Comment (lcnr) [reply compiler/rustc_type_ir/src/search_graph/mod.rs:392]: We have to disable a cache entry if it's the root of a cycle and another cycle participant
is already on the stack.

- ABA cycle
- BA :x:

As the root of a cycle can change its result, we have to also be careful with global cache
entries which depend on a cycle, even if they aren't directly involved.

- ABCB cycle
- CB
    - C cycle
    - A :x:

Comment (compiler-errors) [reply compiler/rustc_type_ir/src/search_graph/mod.rs:412]: Something definitely is not clicking. I don't understand how either that, or the comment you left, translates to:

```
if !nested_goals.is_empty() {
                parent.nested_goals.insert(parent.input, UsageKind::Single(PathKind::Coinductive))
```

Comment (compiler-errors) [reply compiler/rustc_type_ir/src/search_graph/mod.rs:412]: I'm missing something, probably a lot of things.

Comment (lcnr) [reply compiler/rustc_type_ir/src/search_graph/mod.rs:413]: did you also see the inline comment i've added?

**Pull request merged**